### PR TITLE
Implement query complexity analysis #324

### PR DESCRIPTION
**Pull request by ne0h12**
#324

### CODE REVIEW

**Comment by ne0h12** [comment]
@pavelnikolov can you review my PR, please?

File: graphql.go [hunk 1]
```diff
@@ -100,6 +101,23 @@ func MaxDepth(n int) SchemaOpt {
 	}
 }
 
+// MaxQueryComplexity specifies the complexity a query.
```
Comment (pavelnikolov) [inline graphql.go:104]: (nit) "... the complexity **of** a query."

Comment (ne0h12) [reply graphql.go]: I fixed. Thanks.

**Comment by pavelnikolov** [comment]
Hi, @ne0h12 
Sorry for the late reply on this PR. Thank you for your effort but I don't feel that the complexity implementation should look like this. Instead after merging #560, I would like it to be possible for users to define their own `@complexity` directive on `FIELD_DEFINITION` and `ARGUMENT_DEFINITION`. This way each user can assign custom complexity and calculate it based on their needs. This library should not impose a particular solution. Instead it should be extendable and allow the users to define this behavior themselves.

File: internal/validation/complexity.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+func (s visitedSels) copy() visitedSels {
+	newSels := visitedSels{}
+	for index, value := range s {
+		newSels[index] = value
+	}
+
+	return newSels
+}
+
+func (e RecursionEstimator) doRecursivelyVisitSelections(
```
Comment (pavelnikolov) [inline internal/validation/complexity.go:86]: How is this different than the normal `MaxDepth` check?

**Pull request closed without merging**